Support Symfony Expression Language in Simple Tokens

Toflar opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | none
| Docs PR or issue | TODO

I've replaced our own parser for our "Simple Tokens" with the Symfony Expression Language which unlocks a whole new range of functions you can use within Simple Tokens. E.g.

* Array functionality (just use `{if value in ["foo", "bar"]}` instead of 400 if comparisons)
* Object functionality (you can pass objects as values and operate on them)
* You can now expand on the condition and use AND and OR (e.g.  `{if value == "foo" || value == "bar"}`
* You can use regular expressions using `matches`

There's more! See https://symfony.com/doc/current/components/expression_language/syntax.html

Toflar committed:
Added support for Symfony expression language in StringUtil::parseSimpleTokens()

Toflar commented on core-bundle/tests/Contao/StringUtilTest.php:
```diff
@@ -308,16 +302,40 @@ public function parseSimpleTokensProvider(): \Generator
+            'match',
         ];
 
-        yield 'Test does not support line breaks in expressions' => [
```
I've removed the tabs tests because the EL can handle that and I don't see why we wouldn't.

Toflar commented on core-bundle/tests/Contao/StringUtilTest.php:
```diff
@@ -218,12 +218,6 @@ public function parseSimpleTokensProvider(): \Generator
             '{if token=="foo"}',
         ];
 
-        yield 'Test nested if-tag with " in value (match)' => [
```
That's the only thing that's technically a BC break. However, I don't know why we ever supported that because it's clearly invalid syntax. I hope that's okay. /cc @ausi

ausi replied:
I’m OK with breaking that.

fritzmg commented on core-bundle/tests/Contao/StringUtilTest.php:
```diff
@@ -218,12 +218,6 @@ public function parseSimpleTokensProvider(): \Generator
         ];
 
-        yield 'Test nested if-tag with " in value (match)' => [
-            '{if value=="f"oo"}1{endif}{if value=="f\"oo"}2{endif}',
```
The test for escaping, i.e.
```
{if value=="f\"oo"}2{endif}
``` 
should stay though :)

Toflar replied:
Added in ea5a60400ecef48318ac824109c09e7db183fc2f.

Toflar committed:
Bring back the escaping tests

ausi commented on core-bundle/src/Resources/contao/library/Contao/StringUtil.php:
```diff
@@ -558,117 +559,55 @@ public static function parseSimpleTokens($strString, $arrData)
+		$evaluateExpression = static function ($strExpression) use ($el, $arrTokenKeyHashes, $arrData)
+		{
+			// Replace with hashes
+			$strExpression = str_replace(array_keys($arrTokenKeyHashes), array_values($arrTokenKeyHashes), $strExpression);
```
I don’t think this can work reliably. E.g. `{if firstname == "myname"}` with the data `['name' => 'myname mylastname', 'firstname' => 'myname']` breaks.

Toflar replied:
I see, I'm not sure how to fix this, though. It's only the special chars in the variables/tokens which are not supported by the EL (e.g. the `#`). Not sure why one would ever use those...

ausi reviewed (commented):
I’m not sure if we can use `ExpressionLanguage` as is here. We should make sure that it is configured so that it can be used with an expression from an untrusted source. Currently it is possible to access constants with the `constant()` function for example, which would be an information disclosure.

Toflar commented:
> I’m not sure if we can use `ExpressionLanguage` as is here. We should make sure that it is configured so that it can be used with an expression from an untrusted source. Currently it is possible to access constants with the `constant()` function for example, which would be an information disclosure.

I was thinking about this too. At first I wanted to turn this into a service so we can inject the EL and extend it by custom functions. Then I was thinking about the trusted input as well and went for the static implemenation for that very reason. However, I did not find any problematic use case with `constant()`. It's useful for e.g. `{if constant(PHP_VERSION) > 7}` and it cannot be used in tokens. So it can only be used in evaluations which means the value cannot technically leak.
But it would be super easy to disable it, it's the only one registered by default.

aschempp reviewed (changes requested):
I very much like the idea. But I think we can keep BC so far as to check the token keys first. If one is _invalid_, we can still run the old method (obviously not supporting new features then).

ausi commented:
> check the token keys first. If one is _invalid_, we can still run the old method

This would be a great approach IMO! 💪😎👍
If all `array_keys($arrData)` are valid variable names we run `$el->evaluate()` otherwise we trigger a deprecation notice and use the old `$evaluateExpression` method. 🎉

Toflar committed:
Automatically fall back to legacy token parsing if incorrect tokens are passed along

Toflar committed:
Make sure the constant() function is disabled

Toflar commented:
I've addressed all comments here. The default function `constant()` is now disabled (it's the only one) and we're falling back to the legacy evaluation if non-php-variable-compatible tokens have been passed. Works like a charm :)

The only thing that remains breaking now, is the `{if value=="f"oo"}1{endif}`. I cannot really check for this because it's not an invalid token but an invalid comparison value instead.
However, I'm personally totally okay with not supporting that anymore. I don't think there's anyone really using that in the wild. It has to be escaped.

ausi commented on core-bundle/src/Resources/contao/library/Contao/StringUtil.php:
```diff
@@ -582,8 +583,43 @@ static function (array $matches) use ($arrData)
+				}
+				catch (\Exception $e)
+				{
+					throw new \InvalidArgumentException($e->getMessage());
```
Why do we catch the exception and throw another one with the same message?
Couldn’t we just remove the try/catch block herer?

Toflar replied:
The method now throws an `\InvalidArgumentException` so I wanted to convert them and do the same in case someone is catching `\InvalidArgumentException` only when calling `StringUtil::parseSimpleTokens()`.

ausi replied:
We should add the parent exception then I think

```suggestion
					throw new \InvalidArgumentException($e->getMessage(), 0, $e);
```

Toflar replied:
Makes sense, done :)

Toflar committed:
Add parent exception

Co-Authored-By: Martin Auswöger <martin@auswoeger.com>

Toflar committed:
Do not instantiate expression language if not needed

aschempp reviewed (changes requested):
The current implementation leaves tokens as-is if they are no in the token array. Can we add a unit test to make sure thats still the case with the Expression Language?

Toflar commented:
What? No, we cannot test that explicitly and it's actually already there. We test if the method returns the correct string and it's only possible if values remain unchanged. There's no way to unit test from the outside what happens inside and there's absolutely no need to anyway.

aschempp commented:
I meant adding this test (which currently passes):

```php
yield 'Test ignores unknown tokens' => [
    "This is my ##token## that ##remains##",
    ['token' => 'value'],
    'This is my value that ##remains##',
];
```

Toflar committed:
Added an additional test

Toflar commented:
Added test in 2490904c339396949357a46d30a75f64f2b1dd77.

Toflar committed:
CS

Toflar commented:
All comments addressed, all green. RTM.

leofeyer committed:
Fix the coding style

leofeyer commented on core-bundle/src/Resources/contao/library/Contao/StringUtil.php:
```diff
@@ -582,8 +583,53 @@ static function (array $matches) use ($arrData)
 		};
 
-		$evaluateExpression = static function ($strExpression) use ($arrData)
+		// Check if we can use the expression language or if legacy tokens have been used
```
Should we add a comment explaining what legacy tokens are? The discussion says "non-php-variable-compatible tokens" but I have no idea what that means TBH. 🙈

Toflar replied:
Well everything that is not compatible with a php variable? You can even read the regex we're using. E.g. you cannot have a php variable that starts with a digit etc.

leofeyer replied:
As if I knew what `\x7f-\xff` means without looking it up. 😄

Toflar replied:
Nobody uses those anyway. It's just what the PHP parser allows to use as variables. That's why it says "php-variable-compatible".

leofeyer commented:
@aschempp Your review from April 7th is blocking the merge. I think your changes have been addressed though.

aschempp commented on core-bundle/tests/Contao/StringUtilTest.php:
```diff
@@ -68,30 +68,6 @@ public function parseSimpleTokensProvider(): \Generator
             'This is my test@foobar.com,foo@test.com',
         ];
 
-        yield 'Test token replacement with special characters (-)' => [
-            'This is my ##e-mail##',
-            ['e-mail' => 'test@foobar.com'],
-            'This is my test@foobar.com',
-        ];
-
-        yield 'Test token replacement with special characters (&)' => [
-            'This is my ##e&mail##',
-            ['e&mail' => 'test@foobar.com'],
-            'This is my test@foobar.com',
-        ];
-
-        yield 'Test token replacement with special characters (#)' => [
-            'This is my ##e#mail##',
-            ['e#mail' => 'test@foobar.com'],
-            'This is my test@foobar.com',
-        ];
-
-        yield 'Test token replacement with token delimiter (##)' => [
-            'This is my ##e##mail##',
-            ['e##mail' => 'test@foobar.com'],
-            'This is my ##e##mail##',
-        ];
-
```
could these tests be reverted now that we restored the original functionality?

Toflar replied:
they are still there...

ausi replied:
@aschempp see line 329 `parseSimpleTokensLegacyProvider()`

aschempp commented on core-bundle/src/Resources/contao/library/Contao/StringUtil.php:
```diff
@@ -582,8 +583,53 @@ static function (array $matches) use ($arrData)
+
+		if ($canUseExpressionLanguage)
+		{
+			$expressionLanguage = new ExpressionLanguage();
```
Is there a reason we don't configure a service for that? Because if I want to parse (modern) simple tokens in a Contao 4 bundle, I could then inject that service instead of using `StringUtil`. Or would it be recommended to use a "normal" ExpressionLanguage instance for that?

Toflar replied:
@m-vo wanted to extract that into a service later on but can we please finally get this merged?

m-vo replied:
Yeah, I think we can do that in a second step without doing any harm.

Toflar replied:
Ping ;)

m-vo replied:
Do we want to have a `SimpleTokenParser` service or just something for the EL bit?

Toflar replied:
Imho the parser service because who knows what other kind of services we're going to need (and it already needs the logger afaics).

m-vo replied:
[Pong](https://github.com/contao/contao/pull/1756) ;-)

leofeyer commented:
Thank you @Toflar.

contaoacademy commented:
Is it possible to use Arithmetic Operators to calculate some result?

For example:
Sum of 2 Fields: `{field1 + field2} `

m-vo commented:
You can use calculations in expressions but they are only used to evaluate if a condition is true/false.

contaoacademy commented:
@m-vo Thanks for your information. It would be awesome, if we could do some simple arithmetic operations. Instead of `{if} `this could be done by the keyword `{calc}`.
Should I open a feature request?

m-vo commented:
Sure, go ahead. :slightly_smiling_face: Maybe also include your use case - perhaps there are other alternatives.
